packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -1471,11 +1438,193 @@ class RenderPhysicalModel extends _RenderCustomClip<RRect> {
+       _textDirection = textDirection,
+       super(
+         child: child,
+         elevation: elevation,
+         color: color,
+         shadowColor: shadowColor,
+         clipper: clipper,
+       );
+
+  /// The text direction to use for getting the outer path for [shape].
+  ///
+  /// [ShapeBorder]s can depend on the text direction (e.g having a "dent"
+  /// towards the start of the shape).
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (textDirection == value)
+      return;
+    _textDirection = value;
+    _markNeedsClip();
+  }
+
+  @override
+  Path get _defaultClip => new Path()..addRect(Offset.zero & size);
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null) {
+      _updateClip();
+      final Rect offsetBounds = offset & size;
+      final Path offsetPath = _clip.shift(offset);
+      if (needsCompositing) {
+        final PhysicalShapeLayer physicalShape = new PhysicalShapeLayer(
+          clipPath: offsetPath,
+          elevation: elevation,
+          color: color,
+        );
+        context.pushLayer(physicalShape, super.paint, offset, childPaintBounds: offsetBounds);
+      } else {
+        final Canvas canvas = context.canvas;
+        print('Clipping $this');
+        if (elevation != 0.0) {
+          // The drawShadow call doesn't add the region of the shadow to the
+          // picture's bounds, so we draw a hardcoded amount of extra space to
+          // account for the maximum potential area of the shadow.
+          // TODO(jsimmons): remove this when Skia does it for us.
+          canvas.drawRect(
+            offsetBounds.inflate(20.0),
+            _RenderPhysicalModelBase._transparentPaint,
+          );
+          canvas.drawShadow(
+            offsetPath,
+            shadowColor,
+            elevation,
+            color.alpha != 0xFF,
+          );
+        }
+        canvas.drawPath(offsetPath, new Paint()..color = color..style = PaintingStyle.fill);
+        canvas.saveLayer(offsetBounds, _RenderPhysicalModelBase._defaultPaint);
+        canvas.clipPath(offsetPath);
+        super.paint(context, offset);
+        canvas.restore();
+        assert(context.canvas == canvas, 'canvas changed even though needsCompositing was false');
+      }
+    }
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder description) {
+    super.debugFillProperties(description);
+    description.add(new DiagnosticsProperty<CustomClipper<Path>>('clipper', clipper));
+    description.add(new DiagnosticsProperty<TextDirection>('textDirection', textDirection));
+  }
+}
+
+/// A physical model layer casts a shadow based on its [elevation].
+///
+/// The concrete implementations [RenderPhysicalModel] and [RenderPhysicalShape]
+/// determine the actual shape of the physical model.
+abstract class _RenderPhysicalModelBase<T> extends _RenderCustomClip<T> {
```

Comment:
please put this class above its subclasses so that reading the file you get to it first

Reply:
Done.
Is that in our style guide?
I typically used to first see the public, higher-level classes first, and the implementation details below...

Reply:
It's not explicitly in our style guide, but it was confusing to read the patch because there was suddenly mention of this class that hadn't been defined yet. :-)